### PR TITLE
nft scanner with flask app

### PR DESCRIPTION
**Pull request by Rishabh42**

### CODE REVIEW

File: scanner/app.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    result = query_database(nft_address,nft_id)
+
+    # Return result as JSON
+    if result:
+        return {'transfer_event_block': result.transfer_event_block, 'nft_address': result.nft_address, 'nft_id': result.nft_id, 'owner_address': result.owner_address}
+    else:
+        return 'NFT record not found'
+   
+
+if __name__ == '__main__':
```
Comment (ZihangChen) [inline scanner/app.py:77]: Do not run an essential part of the script with name == '__main__'. This is a test function. Start scanning should be in the file body.

File: scanner/app.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+import requests
+import os
+import time
+from dotenv import load_dotenv
+import threading
+from nftScanner import main
+import logging
+
+# load environment variables from .env file
+load_dotenv()
```
Comment (ZihangChen) [inline scanner/app.py:17]: Use a .toml file to load the configuration.

File: scanner/app.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+app = Flask(__name__)
+
+# setup SQL Alchemy
+app.config['SQLALCHEMY_DATABASE_URI'] = f'mysql://{db_user}:{db_password}@{db_host}/{db_name}'
+app.config['SQLALCHEMY_TRACK_MODIFICATIONS'] = False
+db = SQLAlchemy(app)
+
+# Define the NFTOwnership model
+class NFTOwnership(db.Model):
```
Comment (ZihangChen) [inline scanner/app.py:32]: Do not create any orm within the app.py; the app.py should only initialize the server.

File: scanner/sample.env
Comment (ZihangChen) [inline scanner/sample.env:1]: Use .toml file instead of .env

File: scanner/nftScanner.py [hunk 4]
```diff
@@ -10,20 +13,30 @@
...
+console = logging.StreamHandler()
+console.setLevel(logging.INFO)
+# add the handler to the root logger
+logging.getLogger('').addHandler(console)
 
 polygon_url = config('POLYGON_URL')
 
+Base = declarative_base()
+
+class NFTData(Base):
```
Comment (ZihangChen) [inline scanner/nftScanner.py:29]: Create ORM in a separate `model` folder.

File: scanner/nftScanner.py [hunk 5]
```diff
@@ -32,71 +45,109 @@ def __init__(self, cf_contract_address, so_contract_address, from_block):
...
-        self.batch_size = 1000
 
         # DB connection
-        self.mydb = mysql.connector.connect(
-            host="localhost",
-            user=config('DB_USER'),
-            password=config('DB_PASSWORD'),
-            database='nft_data'
-        )
+        self.engine = create_engine('mysql+mysqlconnector://' + config('DB_USER') + ':' + config('DB_PASSWORD') + '@localhost/nft_data')
```
Comment (ZihangChen) [inline scanner/nftScanner.py:50]: The engine should be created when initializing the service. The service should share a connection pool. It is also duplicated with `db = SQLAlchemy(app)` in app.py. Also, the connection loop size should be limited.

File: scanner/nftScanner.py [hunk 6]
```diff
@@ -32,71 +45,109 @@ def __init__(self, cf_contract_address, so_contract_address, from_block):
...
+        self.batch_size = 1000
 
-        # Update owner command
-        self.update_owner_command = 'UPDATE nft_ownership SET transfer_event_block = (%s), owner_address = (%s) WHERE nft_address = (%s) AND nft_ID=(%s)'
-        # Is NFT exists check
-        self.is_nft_exists_command = 'SELECT * from nft_ownership WHERE nft_address = (%s) AND nft_ID=(%s)'
+    # Function to update the ownership of the NFT or insert a newly minted NFT
+    def update_or_insert_nft(self, blockNumber, to, contract_address, token_id):
+        try:
+            Session = sessionmaker(bind=self.engine)
```
Comment (ZihangChen) [inline scanner/nftScanner.py:78]: Duplicated code `Session = sessionmaker(bind=self.engine)`, can create a helper function.

File: scanner/nftScanner.py [hunk 7]
```diff
@@ -32,71 +45,109 @@ def __init__(self, cf_contract_address, so_contract_address, from_block):
...
+            self.session = Session()
```
Comment (ZihangChen) [inline scanner/nftScanner.py:79]: session does not need to be bonded with the class, it should be independently created and used within the function and closed with. So no need to use `self.session`.

File: scanner/app.py [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+def query_database(nft_address, nft_id):
+    # Query database for NFT ownership record
+    try:
+        result = NFTOwnership.query.filter_by(nft_address=nft_address, nft_id=nft_id).first()
+    except Exception as e:
+        logging.error("Error fetching NFT details: ",e)
+
+    return result
+
+@app.route('/get_nft_details', methods=['GET'])
```
Comment (ZihangChen) [inline scanner/app.py:62]: The APIs should be located in a separate folder and file, for code structure management.

File: scanner/app.py [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+    while True:
+        try:
+            main()
+        except Exception as e:
+            logging.error(e)
+        
+        # Delay for 3 seconds
+        time.sleep(3)
+
+def query_database(nft_address, nft_id):
```
Comment (ZihangChen) [inline scanner/app.py:53]: Same with the API, this function should not be in the service initialization.

File: scanner/nftScanner.py [hunk 10]
```diff
@@ -140,16 +187,30 @@ def start_NFT_scan(self, target_block):
             if(blockDiff < self.batch_size):
                 batchSize = blockDiff
 
+        # Update last scanned block
+        try:
+            self.session.query(NFTData).update({'last_scan_block': target_block})
```
Comment (ZihangChen) [inline scanner/nftScanner.py:192]: I think this code is updating every single last_scan_block column within the DB. Should not update all columns without any specification. There should probably only have 1 last_scan_block value in the database.

File: scanner/db/NFT_SQL_COMMANDS.sql [hunk 11]
```diff
@@ -21,18 +21,17 @@ USE `nft_data` ;
 -- Table `nft_data`.`nft_ownership`
 -- -----------------------------------------------------
 CREATE TABLE IF NOT EXISTS `nft_data`.`nft_ownership` (
-  `transfer_event_block` INT NOT NULL,
-  `nft_address` VARCHAR(100) NOT NULL,
-  `nft_ID` INT NULL DEFAULT NULL,
-  `owner_address` VARCHAR(100) NOT NULL)
+  `id` INT PRIMARY KEY NOT NULL AUTO_INCREMENT,
+  `last_scan_block` INT NOT NULL,
```
Comment (ZihangChen) [inline scanner/db/NFT_SQL_COMMANDS.sql:25]: last_scan_block might need to be stored in a separate table.

**Review by ZihangChen — CHANGES_REQUESTED** [review]

File: scanner/nftScanner.py (hunk 10 above)
Comment (Rishabh42) [reply scanner/nftScanner.py:192]: I wasn't sure of the DB design for the NFTs so just added it as a column for now.
But I think I will create a specific table for this

File: scanner/app.py (hunk 8 above)
Comment (Rishabh42) [reply scanner/app.py:62]: As discussed with @ZihangChen , not needed any more

**Pull request merged**